crawl/crawl.go

```diff
@@ -1,178 +1,218 @@
+	index          int64
+	merkleLeafHash []byte
 }
 
 func (ed entryData) Before(other entryData) bool {
 	return ed.index < other.index
 }
 
-func processEntry(prev, curr entryData) {
-	if prev.entry.Leaf.TimestampedEntry.Timestamp > 1000+curr.entry.Leaf.TimestampedEntry.Timestamp {
-		fmt.Println("Found out-of-order entry:")
-		fmt.Printf("  Index: %d\n", prev.index)
-		fmt.Printf("  Timestamps: %d, %d\n", prev.entry.Leaf.TimestampedEntry.Timestamp, curr.entry.Leaf.TimestampedEntry.Timestamp)
-		switch prev.entry.Leaf.TimestampedEntry.EntryType {
-		case ct.X509LogEntryType:
-			cert, err := prev.entry.Leaf.X509Certificate()
-			if err != nil {
-				fmt.Printf("  Failed to parse: %v\n", err)
-			} else {
-				fmt.Printf("  Serial: %d\n", cert.SerialNumber)
+func b64(h []byte) string {
+	return base64.StdEncoding.EncodeToString(h)
+}
+
+func runFetcher(fetcher *scanner.Fetcher) <-chan entryData {
+	entries := make(chan entryData)
+	go func() {
+		err := fetcher.Run(context.Background(), func(batch scanner.EntryBatch) {
+			for i, e := range batch.Entries {
+				h := sha256.New()
+				h.Write([]byte{0})
+				h.Write(e.LeafInput)
+				entries <- entryData{index: batch.Start + int64(i), merkleLeafHash: h.Sum(nil)}
 			}
-		case ct.PrecertLogEntryType:
-			cert, err := prev.entry.Leaf.Precertificate()
-			if err != nil {
-				fmt.Printf("  Failed to parse: %v\n", err)
-			} else {
-				fmt.Printf("  Serial: %d\n", cert.SerialNumber)
+		})
+		if err != nil {
+			log.Fatal(err)
+		}
+		close(entries)
+	}()
+	return entries
+}
+
+// sorted starts a goroutine that reads entryData from its input, and emits
+// them in sorted order (by increasing index) on the output channel it returns.
+func sorter(entries <-chan entryData) <-chan entryData {
+	sortedEntries := make(chan entryData)
+	buffer := NewHeap[entryData]()
+	go func() {
+		var nextIndex int64 = 0
+		for e := range entries {
+			buffer.Push(e)
+
+			for buffer.Len() > 0 && buffer.Peek().index == nextIndex {
+				e = buffer.Pop()
+				sortedEntries <- e
+				nextIndex += 1
 			}
 		}
+		close(sortedEntries)
+	}()
+	return sortedEntries
+}
+
+// verify requests `get-proof-by-hash` for a given entry and calculates the
+// root hash based on the returned audit_path plus the entry's hash. It then
+// compares that against a currently-calculated root hash. This provides a way
+// to do partial consistency checks along the way to verifying a full root hash
+// from scratch. It also ensures that if the log is serving get-proof-by-hash
+// requests using precalculated hashes, those precalculated hashes have not
+// been corrupted.
+func verify(logClient *client.LogClient, e entryData, rootHash []byte) error {
+	currentTreeSize := e.index + 1
+	pbh, err := logClient.GetProofByHash(context.Background(), e.merkleLeafHash, uint64(currentTreeSize))
+	if err != nil {
+		return fmt.Errorf("getting proof by hash: %w", err)
 	}
+
+	hasher := rfc6962.DefaultHasher
+	verifier := logverifier.New(hasher)
```

Comment:
nit, no need to create a local variable that is only used once:
```suggestion
	verifier := logverifier.New(rfc6962.DefaultHasher)
```